GetIntoTeachingApiTests/Controllers/TeachingEventsControllerTests.cs

```diff
@@ -191,6 +192,51 @@ public void ExchangeAccessTokenForAttendee_MissingCandidate_RespondsWithUnauthor
             response.Should().BeOfType<UnauthorizedResult>();
         }
 
+        [Fact]
+        public async Task AddTeachingEvent_WhenRequestIsInvalid_RepondsWithValidationErrorAsync()
+        {
+            const string expectedErrorKey = "Name";
+            const string expectedErrorMessage = "Name must be specified";
+            var request = new TeachingEvent();
+            _controller.ModelState.AddModelError(expectedErrorKey, expectedErrorMessage);
+
+            var response = await _controller.AddOrUpdateTeachingEventAsync(request);
+
+            var badRequest = response.Should().BeOfType<BadRequestObjectResult>().Subject;
+            var errors = badRequest.Value.Should().BeOfType<SerializableError>().Subject;
+            errors.Should().ContainKey(expectedErrorKey).WhichValue.Should().BeOfType<string[]>().Which.Should().Contain(expectedErrorMessage);
+        }
+
+        [Fact]
+        public async Task AddTeachingEvent_WhenRequestIsValid_SavesInCrmAndReturnsCreatedEvent()

```

Comment:
I would probably just have two 'valid' specs here, one to cover saving with a building and one saving without as that's the branch condition in your code (both can check the returned object is correct etc):

```
AddTeachingEvent_ValidRequestWithBuilding_SavesInCrmAndCaches()
AddTeachingEvent_ValidRequestWithoutBuilding_SavesInCrmAndCaches()
```